pkg/eval/external_cmd_test.go

```diff
@@ -0,0 +1,41 @@
+package eval
+
+import (
+	"os"
+	"testing"
+)
+
+func TestBuiltinFnExternal(t *testing.T) {
+	tmpHome, cleanup := InTempHome()
+	defer cleanup()
+
+	mustMkdirAll("dir")
+	mustCreateEmpty("cmdx")
+	mustWriteFile("cmd1", []byte("#!/bin/sh"), 0755)
+	mustWriteFile("cmd2", []byte("#!/bin/sh"), 0755)
+	mustWriteFile("cmd3", []byte(""), 0755)
+	mustCreateEmpty("file")
+
+	os.Setenv("PATH", tmpHome+":"+os.Getenv("PATH"))
+	Test(t,
+		That(`resolve external`).Puts("$external~"),
+		That(`e = (external true); kind-of $e`).Puts("fn"),
+		That(`e = (external true); put (repr $e)`).Puts("<external true>"),
+		That(`e = (external false); m = [&$e=true]; put (repr $m)`).Puts("[&<external false>=true]"),
+		// This group tests the `ExternalCmd.Call` method.
+		That(`e = (external true); $e`).DoesNothing(),
+		That(`e = (external true); $e &option`).Throws(ErrExternalCmdOpts, "$e &option"),
+		That(`e = (external false); $e`).ThrowsCmdExit(
+			ExternalCmdExit{CmdName: "false", WaitStatus: ExitWaitStatus(1)}),
+		// TODO: Modify the ExternalCmd.Call method to wrap the Go error in a
+		// predictable Elvish error so we don't have to resort to using
+		// ThrowsAny in the following tests.
+		//
+		// The command shouldn't be found when run so we should get an
+		// exception along the lines of "executable file not found in $PATH".
+		That(`e = (external true); E:PATH=/; $e`).ThrowsAny(),
+		// The command will be found but cause an exception along the lines of
+		// "exec format error" because it is marked executable but is empty.
+		That(`e = (external cmd3); E:PATH=/; $e`).ThrowsAny(),
```

Comment:
I don't think this test case tests what is claimed; it throws an exception for the same reason the previous test case does.

Reply:
It throws a subtly different error. Whether the two cases should be distinguishable is an open question. See the preceding "TODO". You can argue that they should throw the same exception but they don't which is why I have these two discrete tests to guard against future regressions from the current behavior. Granted, it currently doesn't do so but that's because Elvish doesn't currently provide a predictable exception for each case. Which is why these two tests use `ThrowsAny`.